various fixes

ruslandoga opened the pull request.

ruslandoga commented on config/config.exs:
```diff
@@ -1,3 +1,3 @@
 use Mix.Config
 
-config :clickhousex, codec: Clickhousex.Codec.RowBinary
+# config :clickhousex, codec: Clickhousex.Codec.RowBinary
```
I didn't see `plausible/analytics` setting the codec, so to avoid problems with it, I'm removing this setting as in [the fork](https://github.com/plausible/clickhousex/blob/e6b7cc5a1ea7bb26d77af8fdae03952f662469af/config/config.exs#L3).

ruslandoga commented on lib/clickhousex/http_client.ex:
```diff
@@ -97,13 +97,18 @@ defmodule Clickhousex.HTTPClient do
         {:error, conn, error}
+
+      error ->
+        error
```
`Mint.HTTP.request(conn, "POST", path, headers, post_body)` above can return [`{:error, conn, reason}`.](https://github.com/elixir-mint/mint/blob/9e76b4990f6e0bdc1d1bc7dc5285e65d5ff3af73/lib/mint/http1.ex#L228)

ruslandoga commented on lib/clickhousex/http_client.ex:
```diff
@@ -97,13 +97,18 @@ defmodule Clickhousex.HTTPClient do
   defp decode_response(conn, %Query{type: :select}, %Response{} = response) do
     case Response.decode(response) do
       {:ok, %{column_names: columns, rows: rows}} -> {:ok, conn, {:selected, columns, rows}}
-      {:error, error} -> {:error, conn, error.reason}
```
Not all errors come in a map with `:reason` set. The author seemed to be thinking that `error` here is a mint error struct with `:reason` field set: [`http_error.ex`](https://github.com/elixir-mint/mint/blob/9e76b4990f6e0bdc1d1bc7dc5285e65d5ff3af73/lib/mint/http_error.ex#L60) [`transport_error`](https://github.com/elixir-mint/mint/blob/9e76b4990f6e0bdc1d1bc7dc5285e65d5ff3af73/lib/mint/transport_error.ex#L60). But here the error is [a string, always.](https://github.com/plausible/clickhousex/blob/f0e0820fe79f7e4fa433497bf41fa2e910e51eb0/lib/clickhousex/http_client.ex#L22)


Calling `error.reason`(with error being a string) results in errors being misrepresented as `:erlang.apply`

```
** (ArgumentError) you attempted to apply a function named :reason on "Code: 48. DB::Exception: multi-search all indices is not implemented when vectorscan is off: while executing 'FUNCTION multiMatchAllIndices(pathname : 10, ['^/register$'] : 15) -> multiMatchAllIndices(pathname, ['^/register$']) Array(UInt64) : 16'. (NOT_IMPLEMENTED) (version 22.7.1.2163 (official build))\n". If you are using Kernel.apply/3, make sure the module is an atom. If you are using the dot syntax, such as map.field or module.function(), make sure the left side of the dot is an atom or a map
     stacktrace:
       :erlang.apply("Code: 48. DB::Exception: multi-search all indices is not implemented when vectorscan is off: while executing 'FUNCTION multiMatchAllIndices(pathname : 10, ['^/register$'] : 15) -> multiMatchAllIndices(pathname, ['^/register$']) Array(UInt64) : 16'. (NOT_IMPLEMENTED) (version 22.7.1.2163 (official build))\n", :reason, [])
       (clickhousex 0.5.0) lib/clickhousex/http_client.ex:112: Clickhousex.HTTPClient.decode_response/3
```

instead of the true clickhouse reason

```
** (Clickhousex.Error) Code: 48. DB::Exception: multi-search all indices is not implemented when vectorscan is off: while executing 'FUNCTION multiMatchAllIndices(pathname : 10, ['^/register$', '^/reg[^/]*$', '^/[^/]*/register$', '^/billing.*/success$', '^/billing[^/]*/success$', '^/signup$', '^/signup/[^/]*$', '^/signup/.*$', '^/[^/]*$', '^/.*$'] : 24) -> multiMatchAllIndices(pathname, ['^/register$', '^/reg[^/]*$', '^/[^/]*/register$', '^/billing.*/success$', '^/billing[^/]*/success$', '^/signup$', '^/signup/[^/]*$', '^/signup/.*$', '^/[^/]*$', '^/.*$']) Array(UInt64) : 25'. (NOT_IMPLEMENTED) (version 22.7.1.2163 (official build))
```

ruslandoga commented on lib/clickhousex/http_client.ex:
```diff
@@ -97,13 +97,18 @@ defmodule Clickhousex.HTTPClient do
     case Response.decode(response) do
       {:ok, %{column_names: columns, rows: rows}} -> {:ok, conn, {:selected, columns, rows}}
-      {:error, error} -> {:error, conn, error.reason}
+      {:error, %_exception{} = exception} -> {:error, conn, Exception.message(exception)}
```
Instead, if it's a struct, I assume exception.

ruslandoga commented on lib/clickhousex/http_client.ex:
```diff
@@ -97,13 +97,18 @@ defmodule Clickhousex.HTTPClient do
       {:ok, %{column_names: columns, rows: rows}} -> {:ok, conn, {:selected, columns, rows}}
-      {:error, error} -> {:error, conn, error.reason}
+      {:error, %_exception{} = exception} -> {:error, conn, Exception.message(exception)}
+      {:error, reason} when is_binary(reason) -> {:error, conn, reason}
```
If there is a formatted reason already, it's copied to response.

ruslandoga commented on lib/clickhousex/http_client.ex:
```diff
@@ -97,13 +97,18 @@ defmodule Clickhousex.HTTPClient do
-      {:error, error} -> {:error, conn, error.reason}
+      {:error, %_exception{} = exception} -> {:error, conn, Exception.message(exception)}
+      {:error, reason} when is_binary(reason) -> {:error, conn, reason}
+      {:error, %{reason: reason}} -> {:error, conn, reason}
```
And falling back to what was before.

ruslandoga commented on lib/clickhousex/query.ex:
```diff
@@ -89,10 +89,12 @@ defimpl DBConnection.Query, for: Clickhousex.Query do
+      Regex.match?(@select_query_regex, statement) -> :select
       Regex.match?(@insert_query_regex, statement) -> :insert
       Regex.match?(@create_query_regex, statement) -> :create
-      Regex.match?(@select_query_regex, statement) -> :select
```
This was the reason for https://github.com/plausible/analytics/actions/runs/3204425998/jobs/5235698705 fail.

The [query in the failing test](https://github.com/plausible/analytics/blob/aadf892a71f6db4f7f1e0d961bfc4d7296fa2b59/test/plausible/google/vcr_test.exs#L9-L17) had "Insert" in the `where` clause which made it be classified under ":insert" which resulted in skipped [`FORMAT JSONCompact`](https://github.com/plausible/clickhousex/blob/f0e0820fe79f7e4fa433497bf41fa2e910e51eb0/lib/clickhousex/http_client.ex#L157-L159) and clickhouse responding in tsv (`"3077\n\n"`) and [the match fail.](https://github.com/plausible/clickhousex/blob/f0e0820fe79f7e4fa433497bf41fa2e910e51eb0/lib/clickhousex/codec/json.ex#L42)

ruslandoga replied:
This is now more similar [to current `fork/master`:](https://github.com/plausible/clickhousex/blob/e6b7cc5a1ea7bb26d77af8fdae03952f662469af/lib/clickhousex/query.ex#L87-L95)

```elixir
  defp query_type(statement) do
    cond do
      String.starts_with?(statement, "INSERT") -> :insert
      Regex.match?(@select_query_regex, statement) -> :select
      Regex.match?(@insert_query_regex, statement) -> :insert
      Regex.match?(@alter_query_regex, statement) -> :alter
      true -> :update
    end
  end
```

ruslandoga replied:
(I don't know why `:create` was dropped)

ukutaht reviewed (commented):
Thanks @ruslandoga! Both changes look good to me.

I feel uneasy about the lack of test coverage in this project. Instead of comments like:

> `INSERT` comes first for `INSERT INTO ... SELECT FROM ...` to be marked as insert

Could you add some tests to make sure we have executable documentation for why this specific order of cases is important?

I have the same comment about `error.reason`. There has to be a way to test this properly.

We have recently started using Mox for mocking out things like Finch / Mint in test. That way you can also test errors and responses from the HTTP lib.

I haven't seen you work with unit testing so far. What's your experience with testing? We take it very seriously and happy to help with improving test coverage on this project.

ruslandoga commented on test/clickhousex/query_parse_test.exs:
```diff
@@ -0,0 +1,35 @@
+    test "other statements are merked as `:update`" do
+      assert %Query{type: :update} = Query.new("SHOW DATABASES")
+    end
+  end
```
I've added some basic tests here. I'll look into how `query.type` is actually used to make them more useful.

ukutaht replied:
yeah I'm not sure why `query.type` is significant...

ruslandoga commented on lib/clickhousex/protocol.ex:
```diff
@@ -18,7 +18,7 @@ defmodule Clickhousex.Protocol do
   @type result :: Clickhousex.Result.t()
   @type cursor :: any
 
-  @ping_query Clickhousex.Query.new("SELECT 1") |> DBConnection.Query.parse([])
```
`Clickhousex.Query.new` already calls `DBConnection.Query.parse(statement, [])`, calling it second time is no-op.

ruslandoga commented on mix.exs:
```diff
@@ -33,7 +33,8 @@ defmodule Clickhousex.Mixfile do
       {:benchee, "~> 1.0", only: [:dev, :test]},
-      {:credo, "~> 1.5", only: :dev}
+      {:credo, "~> 1.5", only: :dev},
+      {:bypass, "~> 1.0", only: [:test]}
```
I thought about how to test the http client, and I couldn't come up with a good way to mock it (e.g. there is no behaviour). So instead I'm using bypass to mock the server.

ukutaht replied:
I think bypass is a great solution here

ruslandoga commented on test/clickhousex/http_client_test.exs:
```diff
@@ -0,0 +1,138 @@
+               )
+
+      assert result == {:selected, ["1"], [[1]]}
+    end
```
@ukutaht These are more like drafts right now, just wanted to get feedback before I continue adding more tests.

ruslandoga replied:
I plan to cover every major branch in the http_client.ex logic. And then maybe refactor to make the logic under test more apparent. Like:

```elixir
test "when server responds with 200 to `SELECT 1`", %{conn: conn, bypass: bypass} do
  bottled_response(bypass, :no_db)
  assert {:error, _conn, error} = request(conn, "SELECT 1", database: "no_db")
  assert error == "Code: 81. DB::Exception: Database no_db doesn't exist. (UNKNOWN_DATABASE) ( ..."
end
```

ukutaht replied:
Plan looks good to me!

ukutaht reviewed (commented):
Tests look good! There are some compiler warnings to clean up and more branches to cover but the direction is great.

The main thing is to add tests for the changes made in the PR. So if you revert one of the changes in the PR, at least 1 test should be red. I don't think we need to add exhaustive tests for the whole project in this PR. It certainly would help going forward but at some point you have to make a pragmatic decision and move on. What I'm mostly looking for is to make some quality improvements (testing, refactoring) with each PR in an incremental fashion.

ruslandoga commented:
> There are some compiler warnings to clean up

The warnings come from bypass and existing codebase. I don't think I introduced any new in-app warnings with these changes / tests.

> The main thing is to add tests for the changes made in the PR. So if you revert one of the changes in the PR, at least 1 test should be red.

I've added a test for each change and removed unrelated ones. I'll open a new PR with more tests for changes in `rebase-upstream`.
